### PR TITLE
Precursor map refactor

### PR DESCRIPTION
**Pull request by zachhardy**
This PR is a follow-up to PR #134  addressing issue #133. The precursor_map vector in the TransportCrossSections class was used to store globally defined precursor IDs. This vector was populated within the EigenvalueSolver class. To make the TransportCrossSections class self-contained, this attribute was removed and reformulated as a vector of vectors within the KEigenvalueSolver class. Now, the first index corresponds to the XS ID and the second corresponds to the local precursor ID. This is used within the initialize_precursors routine in order to identify the correct DoF to populate in the precursor vector.

### CODE REVIEW

**Review by ragusa — APPROVED** [review]

File: ChiModules/KEigenvalueSolver/k_eigenvalue_solver.h [hunk 1]
```diff
@@ -16,7 +16,8 @@ class Solver : public LinearBoltzmann::Solver
   size_t source_event_tag;
 
 public:
-  int num_precursors = 0;
+  int num_precursors;
```
Comment (Naktakala) [inline ChiModules/KEigenvalueSolver/k_eigenvalue_solver.h]: Please change `int num_precursors;` to `size_t num_precursors;`.

File: ChiModules/KEigenvalueSolver/initialize_k_solver.cc [hunk 2]
```diff
@@ -27,17 +29,22 @@ void KEigenvalue::Solver::InitializeKSolver()
...
-        }
-
-        // Increment the total number of precursors
-        num_precursors += material_xs[x]->num_precursors;
-      }
+      // Material precursor mapping vector
+      std::vector<size_t> mat_map;
+
+      // Define the precursor mapping for this material
+      for (int j = 0; j < material_xs[x]->num_precursors; ++j)
```
Comment (Naktakala) [inline ChiModules/KEigenvalueSolver/initialize_k_solver.cc]: Please replace `for (int j = 0; ...` with `for (size_t j=0; ...`.

File: ChiModules/KEigenvalueSolver/initialize_k_solver.cc [hunk 3]
```diff
@@ -27,17 +29,22 @@ void KEigenvalue::Solver::InitializeKSolver()
      //       precursor IDs 0-5 and material 1 will receive
      //       precursor IDs 6-11.
     **/
-    num_precursors = 0;
+    precursor_map.clear();
+
+    // Loop over materials
     for (int x = 0; x < material_xs.size(); x++) {
```
Comment (Naktakala) [inline ChiModules/KEigenvalueSolver/initialize_k_solver.cc]: Clang-tidy tells me this loop can be range based. Please change to the suggested below:

```c++
for (auto & xs : material_xs) {
  // Material precursor mapping vector
  std::vector<size_t> mat_map;

  // Define the precursor mapping for this material
  for (int j = 0; j < xs->num_precursors; ++j)
     mat_map.emplace_back(num_precursors + j);

  // Increment the total number of precursors
  num_precursors += xs->num_precursors;

  // Add mapping to the precursor map
  precursor_map.emplace_back(mat_map);
}
```

File: ChiModules/KEigenvalueSolver/initialize_precursors.cc [hunk 4]
```diff
@@ -23,7 +23,8 @@ void KEigenvalue::Solver::InitializePrecursors()
   for (auto& cell : grid->local_cells)
   {
     // ----- Cell information
-    auto xs = material_xs[matid_to_xs_map[cell.material_id]];
+    const auto xs_id = matid_to_xs_map[cell.material_id];
+    auto xs = material_xs[xs_id];
```
Comment (Naktakala) [inline ChiModules/KEigenvalueSolver/initialize_precursors.cc]: This is a silly one but if you "aim to please" then ... you know.

`auto xs` makes a copy of the smart-pointer. Rather use `auto& xs` to use the pointer by reference.

File: ChiModules/KEigenvalueSolver/initialize_precursors.cc [hunk 5]
```diff
@@ -40,7 +41,7 @@ void KEigenvalue::Solver::InitializePrecursors()
         // ----- Loop over precursors
         for (int j = 0; j < xs->num_precursors; ++j)
         {
-          int j_map = xs->precursor_map[j];
+          int j_map = precursor_map[xs_id][j];
```
Comment (Naktakala) [inline ChiModules/KEigenvalueSolver/initialize_precursors.cc]: This is a narrowing conversion. Please change `int j_map` to `size_t j_map`.

**Review by Naktakala — CHANGES_REQUESTED** [review]
Hoping it would merge

File: ChiModules/KEigenvalueSolver/k_eigenvalue_solver.h [hunk 6]
```diff
@@ -16,7 +16,8 @@ class Solver : public LinearBoltzmann::Solver
   size_t source_event_tag;
 
 public:
```
Comment (dominiccaron) [inline ChiModules/KEigenvalueSolver/k_eigenvalue_solver.h:18]: Please document these data members, especially the interpretation/indexing of ```precursor_map```. The text in the description of the pull request is clear, but it is less accessible than if it were directly in the source code.

**Review by dominiccaron — CHANGES_REQUESTED** [review]
Only one request: please add documentation for ```precursor_map``` to the source code.

File: ChiModules/KEigenvalueSolver/k_eigenvalue_solver.h (hunk 1 above)
Comment (zachhardy) [reply ChiModules/KEigenvalueSolver/k_eigenvalue_solver.h]: Done.

File: ChiModules/KEigenvalueSolver/initialize_k_solver.cc (hunk 2 above)
Comment (zachhardy) [reply ChiModules/KEigenvalueSolver/initialize_k_solver.cc]: Done.

File: ChiModules/KEigenvalueSolver/initialize_k_solver.cc (hunk 3 above)
Comment (zachhardy) [reply ChiModules/KEigenvalueSolver/initialize_k_solver.cc]: Done.

File: ChiModules/KEigenvalueSolver/initialize_precursors.cc (hunk 4 above)
Comment (zachhardy) [reply ChiModules/KEigenvalueSolver/initialize_precursors.cc]: Done.

File: ChiModules/KEigenvalueSolver/initialize_precursors.cc (hunk 5 above)
Comment (zachhardy) [reply ChiModules/KEigenvalueSolver/initialize_precursors.cc]: Done.

File: ChiModules/KEigenvalueSolver/k_eigenvalue_solver.h (hunk 6 above)
Comment (zachhardy) [reply ChiModules/KEigenvalueSolver/k_eigenvalue_solver.h:18]: Done.

**Review by Naktakala — APPROVED** [review]

**Review by dominiccaron — APPROVED** [review]

**Review by wdhawkins — APPROVED** [review]

**Pull request merged**